util: Avoid buggy std::filesystem:::create_directories() call

hebasto opened the pull request:
Compiled with some libstdc++ versions (e.g., on Ubuntu 20.04) [`std::filesystem:::create_directories()`](https://en.cppreference.com/w/cpp/filesystem/create_directory) call [fails](https://github.com/bitcoin/bitcoin/issues/24257#issue-1123753243) to handle symbol links properly.

No behavior change in comparison to the [pre-20744](https://github.com/bitcoin/bitcoin/commit/c194293883fbb656779102309b2cb3e60889feff) master branch.

Fixes bitcoin/bitcoin#24257.

hebasto committed:
util: Avoid buggy std::filesystem:::create_directories() call

Compiled with some libstdc++ versions (e.g., on Ubuntu 20.04)
std::filesystem:::create_directories() call fails to handle symbol links
properly.

sipa commented on src/util/system.cpp:
```diff
@@ -443,11 +443,10 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
 
-    if (fs::create_directories(path)) {
-        // This is the first run, create wallets subdirectory too
+    fs::create_directories(path / "wallets");
```
Why do wallet-related things here?

nymkappa commented:
May I ask why you've added the sub directory `/wallets` ? While `bitcoind` and `bitcoin-cli` now runs fine without `-datadir` option (following my symbolic link), wallets are not found anymore because of that change.

```
2022-02-05T01:25:45Z Warning: Skipping -wallet path that doesn't exist. Failed to load database path '/home/joris/.bitcoin/wallets/watching'. Path does not exist.
Warning: Skipping -wallet path that doesn't exist. Failed to load database path '/home/joris/.bitcoin/wallets/watching'. Path does not exist.
```

nymkappa commented:
If you create the `/wallets` directory all the time, it will break current setup who don't use it. Now of course I could just move the folder into the `/wallets` sub-directory and it would work fine, but that's probably not the way to go.

See here the wallet directory loading logic https://github.com/bitcoin/bitcoin/blob/master/src/wallet/walletutil.cpp#L25

nymkappa commented:
@hebasto How about something like this instead? https://github.com/bitcoin/bitcoin/pull/24267/files

hebasto commented:
Updated 94fe141d207999483d54b73a6318cc12fb853b46 -> b9c113af754540341d9529532fbadb7525168102 ([pr24266.01](https://github.com/hebasto/bitcoin/commits/pr24266.01) -> [pr24266.02](https://github.com/hebasto/bitcoin/commits/pr24266.02), [diff](https://github.com/hebasto/bitcoin/compare/pr24266.01..pr24266.02)):

- addressed @nymkappa's [comment](https://github.com/bitcoin/bitcoin/pull/24266#issuecomment-1030503427)

nymkappa commented:
> Updated [94fe141](https://github.com/bitcoin/bitcoin/commit/94fe141d207999483d54b73a6318cc12fb853b46) -> [b9c113a](https://github.com/bitcoin/bitcoin/commit/b9c113af754540341d9529532fbadb7525168102) ([pr24266.01](https://github.com/hebasto/bitcoin/commits/pr24266.01) -> [pr24266.02](https://github.com/hebasto/bitcoin/commits/pr24266.02), [diff](https://github.com/hebasto/bitcoin/compare/pr24266.01..pr24266.02)):
> 
> * addressed @nymkappa's [comment](https://github.com/bitcoin/bitcoin/pull/24266#issuecomment-1030503427)

Tested ACK. This fix is superior to https://github.com/bitcoin/bitcoin/pull/24267 as it keeps [pre-20744](https://github.com/bitcoin/bitcoin/commit/c194293883fbb656779102309b2cb3e60889feff) master branch behavior.

Additional perk is that it allows to follow symbolik links with a multiple depth levels, for example this works fine with this PR:
```
lrwxrwxrwx 1 me me 27 Feb  6 17:05 /home/me/.bitcoin -> /home/me/.bitcoindatadir/
lrwxrwxrwx 1 me me 31 Feb  6 17:04 /home/me/.bitcoindatadir -> /home/me/Documents/.bitcoin//
```

MarcoFalke commented on src/util/system.cpp:
```diff
@@ -443,14 +443,18 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
-        path /= fs::PathFromString(BaseParams().DataDir());
 
-    if (fs::create_directories(path)) {
-        // This is the first run, create wallets subdirectory too
```
why is this comment removed?

hebasto replied:
Code `if (!fs::exists(path)) {` express the same idea, therefore, the removed comment is redundant.

hebasto replied:
@ryanofsky 

Agree. I was thinking about the same some time back and now. And I think it deserves its own PR, keeping this one focused on a workaround for buggy libc++ preserving the current behavior.

ryanofsky reviewed (approved):
Code review ACK b9c113af754540341d9529532fbadb7525168102. Nice simplification and fix

hebasto commented on src/util/system.cpp:
```diff
@@ -443,14 +443,18 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
+        }
+    }
+
     path = StripRedundantLastElementsOfPath(path);
```
> It's harmless and #24265 removes it anyway.

Exactly :)

MarcoFalke commented on src/util/system.cpp:
```diff
@@ -443,14 +443,18 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
 
-    if (fs::create_directories(path)) {
-        // This is the first run, create wallets subdirectory too
+    if (!fs::exists(path)) {
```
Looks like this is also adding a new feature?

Previously, when a test chain was used on the first run, it wouldn't create the wallets dir in the root datadir (`main`). Now it does?

hebasto replied:
On master (8afcc89a8f96692dd880b6102c9a85fac18b9021):
```
$ rm -rf ~/.bitcoin
$ ./src/bitcoind -testnet
$ find ~/.bitcoin -type d
/home/hebasto/.bitcoin
/home/hebasto/.bitcoin/wallets
/home/hebasto/.bitcoin/testnet3
/home/hebasto/.bitcoin/testnet3/wallets
/home/hebasto/.bitcoin/testnet3/chainstate
/home/hebasto/.bitcoin/testnet3/blocks
/home/hebasto/.bitcoin/testnet3/blocks/index
```

MarcoFalke replied:
Ah ok, looks like this is a side effect of reading the config file in the `main` datadir.

maflcko reviewed (approved):
review ACK b9c113af754540341d9529532fbadb7525168102 🐬

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK b9c113af754540341d9529532fbadb7525168102 🐬
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjjYQv/cXNDjyxMuXXzCYEPC7+MwDfQfxws/Xyqbj6nJS4ke7gzmzBfn9VW+pp8
CABVuHSJuux5ElxAJcVWq1ODdlt/qUPeN8Ili7K3i03DvxtlDKyRUfAXneXoNysE
LGVkWd4U2cq4zwidgpjD5yo+QFeXVe9qg00YFxsqevmWQ73o2hJMhJxdAyfVdz8N
QuMZMthT1jvdoBP9ROq+2qNZob4ZBizHWsoHGCOqmyhFJokgXXfUMHZD4TVT321A
zc7M6+daVWH8UTNHxuXw8V9cLDbAMHJhKDj+Y4dOLwFjbtXuD13ZJI1VG5KC4VZo
dN36XSOCsjqpaoNvwQ5dKs28QDNHtAUXw91Dz9vCsYxulY7GrTUtkDxmdGsCjnWX
EcNYGnD+6JagiR9101WG0RbgeOyRnNk12gOhcURX8jtACgCcW6CzQyhvfUtY9Nhj
KBnSJYOIOstzhkDq7BBnw41+3R25ezZmaI8yQs+H0lOw04eTtxMiZYmi0Chg9WwY
LnF0vBuD
=+A1J
-----END PGP SIGNATURE-----
```


</details>

laanwj commented:
I run into a similar issue with the `blocks` directory. I think it would be better to define our own drop in replacement, say, fsbridge::create_directories and use it everywhere that `fs::create_directories` is, not work around this in every single place.
